### PR TITLE
Refactor Qt packages to prepare Qt6

### PR DESCRIPTION
**Pull request by SirLynix**
* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

### CODE REVIEW

**Comment by SirLynix** [comment]
I don't understand, all packages are not tested

**Comment by waruqi** [comment]
> I don't understand, all packages are not tested

please remove it, it will override the kind of qt5core, qt5gui

```diff
package("qt5lib")
    set_kind("template")
    set_homepage("https://www.qt.io")
    set_description("Qt is the faster, smarter way to create innovative devices, modern UIs & applications for multiple screens. Cross-platform software development at its best.")
    set_license("LGPL-3")

    on_load(function (package)
-       if package.is_template then
-           package:set("kind", "template")
-       end
        package:add("deps", "qt5base", {debug = package:is_debug(), version = package:version_str()})
    end)
```

**Comment by SirLynix** [comment]
It seems find_qt can't find the android/ios installation, as it relies on `qmake` (which only exists as a host binary)

**Comment by waruqi** [comment]
> It seems find_qt can't find the android/ios installation, as it relies on `qmake` (which only exists as a host binary)

I tried it on my machine, there is qmake for android.

```
$  aqt install-qt mac android 5.13.2 android_armv7
 /Users/ruki/Downloads/aqt/5.13.2/android_armv7/bin/qmake
```

**Comment by SirLynix** [comment]
you're right, I didn't see it. however it's replaced by host binaries which find the host info

**Comment by waruqi** [comment]
```
arch android
aqt install-qt -O /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92
mac android 5.15.2 android
aqtinstall(aqt) v2.0.5 on Python 3.9.13 [CPython Clang 13.0.0 (clang-1300.0.29.30)]
Downloading qtwebsockets...
Downloading qtbase...
Downloading qtsvg...
Redirected: mirrors.sjtug.sjtu.edu.cn
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtwebsockets-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 2.
54058865
Downloading qttools...
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtdeclarative...
Finished installation of qtsvg-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 3.6101373
4
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtdeclarative-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 5
.79202664
Downloading qtmultimedia...
Finished installation of qtbase-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 13.03684
733
Downloading qtxmlpatterns...
Redirected: mirrors.sjtug.sjtu.edu.cn
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtmultimedia-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 7.
70419455
Downloading qtwebchannel...
Finished installation of qtxmlpatterns-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 4
.35613605
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtandroidextras...
Finished installation of qtwebchannel-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 2.
63691611
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtsensors...
Finished installation of qtandroidextras-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in
 2.57362207
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtsensors-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 3.313
20881
Downloading qtconnectivity...
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtserialport...
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qttranslations...
Finished installation of qtserialport-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 3.
83339222
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtlocation...
Finished installation of qtconnectivity-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in
7.07468992
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qttools-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 27.8241
2246
Finished installation of qttranslations-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in
24.39722360
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtquickcontrols...
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtimageformats...
Finished installation of qtlocation-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 4.44
759730
Downloading qt3d...
Finished installation of qtquickcontrols-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in
 3.51188932
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtgraphicaleffects...
Downloading qtquickcontrols2...
Finished installation of qtimageformats-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in
3.88808861
Redirected: mirrors.sjtug.sjtu.edu.cn
Redirected: mirrors.sjtug.sjtu.edu.cn
Downloading qtscxml...
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtgraphicaleffects-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z
 in 2.94249039
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtquickcontrols2-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z i
n 3.41973201
Finished installation of qtscxml-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 2.79195
723
Finished installation of qt3d-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 6.83071925
Downloading qtspeech...
Downloading qtgamepad...
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtspeech-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 2.6328
8022
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtgamepad-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 3.006
85563
Downloading qtwebview...
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtwebview-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in 4.418
75361
Downloading qtremoteobjects...
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtremoteobjects-MacOS-MacOS_10_13-Clang-Android-Android_ANY-Multi.7z in
 6.82486593
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/5.15.2/an
droid/bin/qmake
Finished installation
Time elapsed: 45.63361180 second
aqt install-qt -O /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/
bin_host mac desktop 5.15.2 --archives qtbase
aqtinstall(aqt) v2.0.5 on Python 3.9.13 [CPython Clang 13.0.0 (clang-1300.0.29.30)]
Downloading qtbase...
Redirected: mirrors.sjtug.sjtu.edu.cn
Finished installation of qtbase-MacOS-MacOS_10_13-Clang-MacOS-MacOS_10_13-X86_64.7z in 9.7706430
2
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/bin/qmake
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Xml.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Network.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5OpenGLExtensions.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Widgets.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5DBus.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5OpenGL.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Core.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Test.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Gui.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Sql.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5PrintSupport.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/pkgconfig/Qt5Concurrent.pc
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5ClipboardSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5GraphicsSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5PlatformCompositorSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5ServiceSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5EdidSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5DeviceDiscoverySupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5EventDispatcherSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5FontDatabaseSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5OpenGLExtensions.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5ThemeSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5FbSupport.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5Bootstrap.la
Patching /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/
5.15.2/clang_64/lib/libQt5AccessibilitySupport.la
Finished installation
Time elapsed: 13.41199624 second
checking for Qt SDK directory ... /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518
319e656e8919e92/bin_host/5.15.2/clang_64
checking for Qt SDK version ... 5.15.2
checking for local-repo::qt5base ... qt5base 5.15.2
{
  pluginsdir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bi
n_host/5.15.2/clang_64/plugins",
  version = "5.15.2",
  sdkver = "5.15.2",
  mkspecsdir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bi
n_host/5.15.2/clang_64/mkspecs",
  qmldir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_ho
st/5.15.2/clang_64/qml",
  includedir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bi
n_host/5.15.2/clang_64/include",
  sdkdir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_ho
st/5.15.2/clang_64",
  libdir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_ho
st/5.15.2/clang_64/lib",
  libexecdir_host = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e
92/bin_host/5.15.2/macos/libexec",
  libexecdir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bi
n_host/5.15.2/clang_64/libexec",
  bindir = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_ho
st/5.15.2/clang_64/bin",
  bindir_host = "/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/b
in_host/5.15.2/clang_64/bin"
}
```
aqt was called twice and found the host qmake

```
aqt install-qt -O /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92
mac android 5.15.2 android
aqt install-qt -O /Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/
bin_host mac desktop 5.15.2 --archives qtbase
```

```
/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin/qmake
/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin_host/5.15.2/clang_64/bin/qmake
```

**Comment by waruqi** [comment]
we should find `/Users/ruki/.xmake/packages/q/qt5base/5.15.2/481e216908744a518319e656e8919e92/bin/qmake`

File: packages/q/qtbase/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            else
+                arch = "android"
+            end
+        end
+
+        local installdir = package:installdir()
+        print("host", host)
+        print("target", target)
+        print("versionstr", versionstr)
+        print("arch", arch)
```
Comment (waruqi) [inline packages/q/qtbase/xmake.lua:144]: please remove logs

File: packages/q/qtbase/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                    moc = true,
+                    rcc = true,
+                    uic = true
+                }
+            end
+
+            for folder, tools in pairs(tool_folders) do
+                for _, file in pairs(os.files(path.join(installdir, "bin_host", versionstr, "*", folder, "*"))) do
+                    local filename = path.filename(file)
+                    if (tools[filename]) then
```
Comment (waruqi) [inline packages/q/qtbase/xmake.lua:190]: remove `()`

File: packages/q/qtbase/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                        local targetpath = path.join(installdir, folder, filename)
+                        os.rm(targetpath)
+                        os.ln(file, targetpath)
```
Comment (waruqi) [inline packages/q/qtbase/xmake.lua:193]: please do not use softlink, it maybe does not work on windws, you can use `os.cp`

File: packages/q/qtbase/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+                        -- some tools like CMake will try to run moc.exe even on Linux, trick them (ln bin/moc.exe => bin_host/bin/moc)
+                        if package:is_plat("mingw") then
+                            os.rm(targetpath .. ".exe")
+                            os.ln(file, targetpath .. ".exe")
```
Comment (waruqi) [inline packages/q/qtbase/xmake.lua:198]: and here.

File: packages/q/qtbase/xmake.lua (hunk 3 above)
Comment (SirLynix) [reply packages/q/qtbase/xmake.lua]: this code should not be executed on windows, but I added a os.cp fallback in case it does in the future.

**Comment by waruqi** [comment]
download error?

**Comment by SirLynix** [comment]
yes, the Qt download mirrors appear to be down for now.

**Comment by SirLynix** [comment]
```
note: the following packages were not found on your system, try again after installing them:
  -> openssl#1 1.1.1-t
```
Why Fedora can't install openssl?

**Comment by waruqi** [comment]
> ```
> note: the following packages were not found on your system, try again after installing them:
>   -> openssl#1 1.1.1-t
> ```
> 
> Why Fedora can't install openssl?

I don't know, I need some time to debug it.

**Comment by waruqi** [comment]
> ```
> note: the following packages were not found on your system, try again after installing them:
>   -> openssl#1 1.1.1-t
> ```
> 
> Why Fedora can't install openssl?

here

```lua
package("qt5network")
    set_base("qt5lib")
    set_kind("library")

    on_load(function (package)
        package:add("deps", "qt5core", {debug = package:is_debug(), version = package:version_str()})
        package:data_set("libname", "Network")

        if package:is_plat("linux") then
            -- we need system openssl with evp-kdf
            -- @see https://github.com/xmake-io/xmake-repo/pull/1057#issuecomment-1069006866
            if linuxos.name() == "fedora" then
                package:add("deps", "openssl", {system = true})     <-------- system
            else
                package:add("deps", "openssl")
            end
```

**Comment by SirLynix** [comment]
Oh right, I forgot about this. So I think we can merge this pull request.

**Pull request merged**